### PR TITLE
HPCC-28434 Add network policy for spray service

### PR DESCRIPTION
**Pull request by jakesmith**
Generate a spray service network policy to permit ingress into the dfuserver(s) that use them.

Signed-off-by: Jake Smith <jake.smith@lexisnexisrisk.com>

## Type of change:
- [x] This change is a bug fix (non-breaking change which fixes an issue).
- [ ] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [ ] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [ ] The commit message title makes sense in a changelog, by itself.
  - [ ] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [ ] The change has been fully tested:
  - [ ] I have added tests to cover my changes.
  - [ ] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [ ] Scalability
  - [ ] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [ ] Cloud-compatibility
  - [ ] Premature optimization
  - [ ] Existing deployed queries will not be broken
  - [ ] This change fixes the problem, not just the symptom
  - [ ] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Smoketest:
- [ ] Send notifications about my Pull Request position in Smoketest queue.
- [ ] Test my draft Pull Request.

### CODE REVIEW

**Comment by jakesmith** [comment]
@rwagner42 - please test

File: helm/hpcc/templates/_helpers.tpl [hunk 1]
```diff
@@ -1011,6 +1011,7 @@ Generate list of available services
 {{- range $dafilesrv := $.Values.dafilesrv -}}
  {{- if not .disabled }}
 - name: {{ .name }}
+  class: dafilesrv
```
Comment (jakesmith) [inline helm/hpcc/templates/_helpers.tpl:1014]: incidental - this doesn't effect anything, but was missing.

File: helm/hpcc/templates/dafilesrv.yaml [hunk 2]
```diff
@@ -50,6 +50,7 @@ spec:
       labels:
         {{- include "hpcc.addStandardLabels" (dict "root" $ "component" "dafilesrv" "name" "dafilesrv" "instance" .name) | indent 8 }}
         server: {{ .name | quote }}
+        helmVersion: 8.10.3-closedown0
```
Comment (jakesmith) [inline helm/hpcc/templates/dafilesrv.yaml:53]: helmVersion was missing in dafilesrv.yaml - this matches the same setting in other template yamls.
It/they are manipulated on a rc and gold release.

**Comment by jakesmith** [comment]
> @rwagner42 - please test

Russell has not confirmed the fix resolves the issue.

**Comment by rwagner3** [comment]
Fixed tested, works.

File: helm/hpcc/templates/dafilesrv.yaml [hunk 3]
```diff
@@ -101,5 +102,45 @@ kind: ConfigMap
 {{ include "hpcc.addService" ( dict "root" $ "name" .name "service" $service "selector" .name "defaultPort" 7600 ) }}
 ---
 {{ include "hpcc.addCertificate" (dict "root" $ "name" .name "service" .service "component" "dafilesrv" "visibility" .service.visibility) }}
+{{- if eq .application "spray" -}}
```
Comment (afishbeck) [inline helm/hpcc/templates/dafilesrv.yaml:105]: A comment above this line marking the start of where the generation of the Network policy may help the readability a bit.

File: helm/hpcc/templates/dafilesrv.yaml [hunk 4]
```diff
@@ -101,5 +102,45 @@ kind: ConfigMap
...
+     {{- $sprayServiceNames := (get $networkPolicyCtx.dfuServerMap $dfuServer.sprayServiceName) | default list -}}
+     {{- $sprayServiceNames := append $sprayServiceNames $dfuServer.name -}}
+     {{- $_ := set $networkPolicyCtx.dfuServerMap $dfuServer.sprayServiceName $sprayServiceNames -}}
+    {{- else -}}
+     {{- $_ := set $networkPolicyCtx "untargetedDfuServers" (append $networkPolicyCtx.untargetedDfuServers $dfuServer.name) -}}
+    {{- end -}}
+   {{- end -}}
+  {{- end -}}
+ {{ end }}
+ {{- $_ := set $commonCtx "allowedDfuServers" (ternary $networkPolicyCtx.untargetedDfuServers list $networkPolicyCtx.firstSprayService) -}}
```
Comment (afishbeck) [inline helm/hpcc/templates/dafilesrv.yaml:121]: A comment above stating that dfu servers without a "sprayServiceName" set are assigned to the first spray service may help here.

**Review by afishbeck — APPROVED** [review]
@jakesmith a couple minor suggestions for code comments, but otherwise looks good.

**Comment by jakesmith** [comment]
@afishbeck - added suggested comments (squashed in)

**Comment by jakesmith** [comment]
@ghalliday - please merge

**Review by richardkchapman — APPROVED** [review]

**Pull request merged**